README
NAME
    Ado::Plugin::Admin - system (site) administration user interface

SYNOPSIS
      #Add this plugin to configuration file
      # /home/you/whereinstalled/Ado/etc/ado.$mode.config
      plugins => {
        # other plugins here...
        'admin',
        # other plugins depending on Ado::Plugin::Admin here...
      }

      #run/restart ado
      $ cd /hereis/ado
      morbo ./bin/ado

      #go to http://localhost:3000/ado
      #enjoy

DESCRIPTION
    Ado::Plugin::Admin provides:

    *   REST API for default home page for the administration area of your
        site or web-application;

    *   API for other plugins that want to provide user interface for
        administering various parts of the application;

    *   Default UI consisting of a left panel and a default dashboard. The
        dashboard consists of descriptions of the installed plugins that
        provide description for the dashboard

        Note: This software is not funtional yet!

METHODS
    Ado::Plugin::Admin inherits all methods from Ado::Plugin and implements
    the following new ones.

  register
      $plugin->register(Ado->new);

    Initializes and registers plugin in Ado application.

HELPERS
    Ado::Plugin::Admin provides the following helpers

  admin_menu
    TODO: Renders the main menu in the administration screen in html or json
    for the current user.

      <%= admin_menu %>

CONTINUOUS INTEGRATION
    We would like to know that our software is always in good health. We
    count on friendly developers and organizations to install and test it
    continuously.

    CPAN Testers Reports for Ado::PLugin::Admin
    <http://www.cpantesters.org/distro/A/Ado-Plugin-Admin.html>

    Travis-CI <https://travis-ci.org/kberov/Ado-Plugin-Admin>

SEE ALSO
    Ado::UI::Menu, Ado::Control::Ado::Default, Ado::Control::Ado::Users,
    Ado::Plugin, Mojolicious::Guides::Growing, Ado::Manual, Mojolicious,
    <http://mojolicio.us>.

AUTHOR
    Krasimir Berov

COPYRIGHT AND LICENSE
    Copyright 2015 Krasimir Berov.

    This program is free software, you can redistribute it and/or modify it
    under the terms of the GNU Lesser General Public License v3 (LGPL-3.0).
    You may copy, distribute and modify the software provided that
    modifications are open source. However, software that includes the
    license may release under a different license.

    See http://opensource.org/licenses/lgpl-3.0.html for more information.



Changes
Revision history for Ado-Plugin-Admin

0.14 2015-04-05
  - Fixed html section in Admin.pm documentation.

0.13 2015-04-05
  - Upgraded to Ado 0.87!
  - Cleaned up MANIFEST. Added MANIFEST.SKIP to MANIFEST.


0.12 2015-03-29
  - Upgraded to Ado 0.86!
  - Added list 'adobar_links' to session and stash for pushing to it
    addtional links for loged-in users and for anyone.
  - Added tests for listing users.
  - Implemented grid for listing users - alpha.
  - Implemented main sidebar - alpha.
  - 2015-03-03 Честит празник на всички българи!
    Днес - трети март - е денят на независимостта
    на третата българска държава.

0.11 2015-02-26
  - Make sure the database file is writable during tests!

0.10 2015-02-25
  - It would be best if the database file was bundled with
    the distribution ;).

0.09 2015-02-25
  - Use existing database file to avoid test falures.

0.08 2015-02-23
  - Trying to fix failing tests by putting the database file
    in the t/plugin/ directory.
    http://www.cpantesters.org/cpan/report/04f49de2-baea-11e4-a6d1-bc568971dd2f

0.07 2015-02-23
  - Trying to fix failing tests by setting 'PRAGMA synchronous = ON'.
    http://www.cpantesters.org/cpan/report/1b8448b4-bae5-11e4-be01-88698971dd2f

0.06 2015-02-22
  - Use plugin's own ado.conf and database for testing.
    http://www.cpantesters.org/cpan/report/09e78bd2-ab6f-11e4-a5b3-6a9cdfbfc7aa
    Solved problem with file access during tests. Thanks Slaven Rezić (SREZIC).
  - Upgraded to Ado 0.80 where the functionality for the above case is
    implemented.
  - Moved Ado::Control::Ado to Ado-Plugin-Admin where it logically belongs.

0.05 2015-02-06
  - Use our own Ado configuration file for testing.
    http://www.cpantesters.org/cpan/report/09e78bd2-ab6f-11e4-a5b3-6a9cdfbfc7aa

0.04 2015-02-04
  - Let Ado find its home during tests of plugins too.

0.03 2015-02-01
  - Depending explicitly on Ado 0.78.

0.02 2015-02-01
  - Improved documentation and test coverage.
  - Moved Ado::Control::Ado::Users and Ado::Control::Ado::Default
    from Ado (core) here.
  - Implemented Ado::UI::Menu.
    Decommissioned Ado::UI::AdminMenu and Ado::UI::MenuItem.

0.01 2015-01-18
  - Released first dummy version.


lib/Ado/Plugin/Admin.pm
package Ado::Plugin::Admin;
use Mojo::Base 'Ado::Plugin';
use Ado::UI::Menu;
use Mojo::Util qw(monkey_patch);
our $VERSION = '0.14';

sub register {
    my ($self, $app, $config) = shift->initialise(@_);
    $self->_init_admin_menu();

    #User specific Admin menu
    $app->helper(admin_menu => \&_render_admin_menu);

    #switch layouts depending on XMLHttpRequest
    $app->hook(
        around_action => sub {
            my ($next, $c, $action, $last_in_stack) = @_;
            if ($c->user->login_name ne 'guest') {

                #Add link to admin area for logged-in users
                List::Util::first(
                    sub { $_->{href} eq '/ado' },
                    @{$c->stash->{adobar_links}}
                  )
                  || push @{$c->stash->{adobar_links}},
                  {icon => 'dashboard', href => '/ado', text => 'Dashboard'};
            }

            if (($c->stash->{controller} // '') =~ m|^ado|) {
                if   ($c->req->is_xhr) { $c->layout('admin_xhr') }
                else                   { $c->layout('admin') }
            }
            return $next->();
        }
    );
    $app->hook(
        after_login => sub {
            push @{$_[0]->session->{adobar_links}},
              {icon => 'dashboard', href => '/ado', text => 'Dashboard'};
        }
    );

    #may be used later
    #$app->plugin("AssetPack");
    #$app->asset('admin.css' => ('/plugins/admin/admin.css'));
    #$app->asset('admin.js'  => ('/plugins/admin/admin.js'));
    return $self;
}

sub _init_admin_menu {
    my $self = shift;
    my $menu =
      Ado::UI::Menu->new(header => 1, icon => 'ado', title => 'Plugins');
    $menu->items(
        {title => 'Dashboard', url => '/ado', header => 1, icon => 'dashboard'},
        {title => 'Content', icon => 'content', header => 1,},
        {title => 'System',  icon => 'setting', header => 1,}
      )->first(sub { $_[0]->title eq 'System' })->items(
        {title => 'Settings', icon => 'settings', url => '/ado-settings'},
        {title => 'Users',    icon => 'users',    url => '/ado-users'},
      );
    $menu->items->first(sub { $_[0]->title eq 'Content' })->items(
        {title => 'Pages', icon => 'browser', url => '/ado-pages'},
        {title => 'Blog',  icon => 'content', url => '/ado-blog'},
    );
    monkey_patch ref($self->app), admin_menu => sub {$menu};
    return;
}


# Rendered once and cached in the user session.
sub _render_admin_menu {
    my $c = shift;

    # return prepared HTML
    return $c->session('admin_menu') if $c->session('admin_menu');

    my $menu = $c->render_to_string(
        template => '/ado/partials/admin_menu',
        item     => $c->app->admin_menu
    );
    $c->session('admin_menu' => $menu);    # session cache
    return $menu;
}
1;

=encoding utf8

=head1 NAME

Ado::Plugin::Admin - system (site) administration user interface

=head1 SYNOPSIS

  #Add this plugin to configuration file
  # /home/you/whereinstalled/Ado/etc/ado.$mode.config
  plugins => {
    # other plugins here...
    'admin',
    # other plugins depending on Ado::Plugin::Admin here...
  }

  #run/restart ado
  $ cd /hereis/ado
  morbo ./bin/ado

  #go to http://localhost:3000/ado
  #enjoy

=head1 DESCRIPTION

L<Ado::Plugin::Admin> provides:

=over

=item * REST API for default home page for the administration area
of your site or web-application;

=item * API for other plugins that want to provide user interface for administering various parts of the application;

=item * Default UI consisting of a left panel and a default dashboard.
The dashboard consists of descriptions of the installed plugins that
provide description for the dashboard

B<Note:> This software is not funtional yet!

=back

=head1 METHODS

L<Ado::Plugin::Admin> inherits all methods from
L<Ado::Plugin> and implements the following new ones.

=head2 register

  $plugin->register(Ado->new);

Initializes and registers plugin in L<Ado> application.

=head1 HELPERS

L<Ado::Plugin::Admin> provides the following helpers

=head2 admin_menu

TODO: Renders the main menu in the administration screen in html or json for the current user.

  <%= admin_menu %>

=head1 CONTINUOUS INTEGRATION

We would like to know that our software is always in good health.
We count on friendly developers and organizations to install and test it continuously.

L<CPAN Testers Reports for Ado::PLugin::Admin|http://www.cpantesters.org/distro/A/Ado-Plugin-Admin.html>

L<Travis-CI|https://travis-ci.org/kberov/Ado-Plugin-Admin> 


=begin html

 <a href="https://travis-ci.org/kberov/Ado-Plugin-Admin"><img 
  src="https://travis-ci.org/kberov/Ado-Plugin-Admin.svg?branch=master"
    ></a><br /><br />

=end html

=head1 SEE ALSO

L<Ado::UI::Menu>, L<Ado::Control::Ado::Default>, L<Ado::Control::Ado::Users>,
L<Ado::Plugin>, L<Mojolicious::Guides::Growing>,
L<Ado::Manual>, L<Mojolicious>,  L<http://mojolicio.us>.

=head1 AUTHOR

Krasimir Berov

=head1 COPYRIGHT AND LICENSE

Copyright 2015 Krasimir Berov.

This program is free software, you can redistribute it and/or
modify it under the terms of the 
GNU Lesser General Public License v3 (LGPL-3.0).
You may copy, distribute and modify the software provided that 
modifications are open source. However, software that includes 
the license may release under a different license.

See http://opensource.org/licenses/lgpl-3.0.html for more information.

=cut
